src/v6_indexer/utils.ts

```diff
@@ -480,4 +460,305 @@ export async function getVaultBalances(
       console.error("Error fetching vault balances:", error);
       throw error;
     }
-  }
\ No newline at end of file
+  }
+
+/**
+ * Gets token decimals from the database
+ */
+async function getTokenDecimals(db: DBConnection, mintAcct: string): Promise<number> {
+  try {
+    const token = await db.select({ decimals: schema.tokens.decimals })
+      .from(schema.tokens)
+      .where(eq(schema.tokens.mintAcct, mintAcct))
+      .limit(1);
+    
+    return token.length > 0 ? token[0].decimals : 0;
+  } catch (error) {
+    logger.warn(`Error fetching decimals for token ${mintAcct}:`, error);
+    return 0;
+  }
+}
+
+export async function insertIfNotExistsMarkets(
+  db: DBConnection,
+  proposalAddr: string,
+  daoAddr: string,
+  passBaseMint: string,
+  passQuoteMint: string,
+  failBaseMint: string,
+  failQuoteMint: string
+): Promise<void> {
+  try {
+    // Get DAO to find base and quote mints for spot market
+    const dao = await db.select()
+      .from(schema.v0_6_daos)
+      .where(eq(schema.v0_6_daos.daoAddr, daoAddr))
+      .limit(1);
+    
+    if (dao.length === 0) {
+      logger.warn(`DAO ${daoAddr} not found when initializing markets`);
+      return;
+    }
+
+    const markets = [
+      {
+        daoAddr,
+        proposalAddr: null, // Spot market has no proposal
+        marketType: 'spot',
+        baseMint: dao[0].baseMintAcct,
+        quoteMint: dao[0].quoteMintAcct,
+      },
+      {
+        daoAddr,
+        proposalAddr,
+        marketType: 'pass',
+        baseMint: passBaseMint,
+        quoteMint: passQuoteMint,
+      },
+      {
+        daoAddr,
+        proposalAddr,
+        marketType: 'fail',
+        baseMint: failBaseMint,
+        quoteMint: failQuoteMint,
+      }
+    ];
+
+    await db.insert(schema.futarchy_markets).values(markets).onConflictDoNothing();
+    logger.info(`Initialized markets for proposal ${proposalAddr}`);
+  } catch (error) {
+    logger.error(`Error initializing markets for proposal ${proposalAddr}:`, error);
+  }
+}
+
+/**
+ * Calculate price from pool reserves with decimal adjustment
+ */
+function calculatePriceWithDecimals(
+  quoteReserves: BN,
+  baseReserves: BN,
+  quoteDecimals: number,
+  baseDecimals: number
+): BN {
+  if (baseReserves.isZero() || quoteReserves.isZero()) {
+    return new BN(0);
```

Comment:
Were' still returning a price which is 0, infinity or undefined... We should bail here.